File: src/common/options/sortev.f90
!
! THIS PROGRAM IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR MODIFY
! IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE AS PUBLISHED BY
! THE FREE SOFTWARE FOUNDATION; EITHER VERSION 2 OF THE LICENSE, OR
! (AT YOUR OPTION) ANY LATER VERSION.
!
! THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
! BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
! MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
! GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
!
! YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE
! ALONG WITH THIS PROGRAM; IF NOT, SEE <HTTP://WWW.GNU.ORG/LICENSES/>.
!
! Copyright (C) Martin Koehler, 2017-2023
!

!> @addtogroup sortev
!> @{

!> @brief Options to sort eigenvalues in the frontend.
!>
!> The MEPACK_OPTIONS_SORTEV module provide a global option to enable
!> or disable the sort of the eigenvalues before a solver
!> is called from the frontend routines.
MODULE MEPACK_OPTIONS_SORTEV
    IMPLICIT NONE

    INTEGER, SAVE, PRIVATE :: SORTEV_ENABLED_VAR = 0

    CONTAINS

        !> @brief Return if the sorting of the eigenvalues is enabled.
        !> @return A non zero value if the eigenvalues should be sorted.
        !>
        !> The SORTEV_ENABLED function returns a non zero value if the eigenvalues
        !> of the coefficient matrices should be sorted before the equations are solved.
        FUNCTION SORTEV_ENABLED() RESULT(SORTEV)
            IMPLICIT NONE
            INTEGER :: SORTEV

            SORTEV = SORTEV_ENABLED_VAR
        END FUNCTION SORTEV_ENABLED

        !> @brief Enable or disable the sorting of the eigenvalues.
        !> @param[in]  SORT         Integer identifying if the eigenvalues should be sorted or not. #
        !>
        !> The SORTEV_ENABLE subroutine enables or disable the sorting of the eigenvalues
        !> before a triangular matrix equation solver is called in a frontend routine.
        !>
        SUBROUTINE SORTEV_ENABLE(SORT)
            IMPLICIT NONE
            INTEGER, INTENT(IN) :: SORT
            SORTEV_ENABLED_VAR = SORT
            RETURN
        END SUBROUTINE SORTEV_ENABLE


END MODULE MEPACK_OPTIONS_SORTEV
!> @}

File: src/common/options/sortev.c
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) Martin Koehler, 2017-2023
 */


#include "FCMangle.h"
#include "mepack.h"
#include "mepack_internal.h"

/**
 * @addtogroup sortev
 * @{
 */

/**
 * @brief Return if the eigenvalue sorting is enabled or not.
 * @return A non-zero if the eigenvalue sorting is enabled.
 *
 * The mepack_sortev_enabled function returns a true value if the eigenvalue
 * sorting is enabled. That means, frontend routines sort the eigenvalues
 * before calling a triangular solver in order to match the block structure.
 */
int mepack_sortev_enabled()
{
    int ret;
    ret = (int)  FC_MODULE_(mepack_options_sortev,sortev_enabled,MEPACK_OPTIONS_SORTEV,SORTEV_ENABLED)();
    return ret;
}

/**
 * @brief Enable or disable the eigenvalues sorting in the frontend routines.
 * @param[in] sort      Boolean integer value to enable or disable the eigenvalue sorting.
 *
 * The mepack_sortev_enable function enables or disables the sorting of the eigenvalues
 * before a triangular solver is called by one of the frontend routines.
 *
 */
void mepack_sortev_enable(int sort)
{
    Int _sort = sort;
    FC_MODULE_(mepack_options_sortev,sortev_enable,MEPACK_OPTIONS_SORTEV,SORTEV_ENABLE)(&_sort);
    return;
}


/**
 * @}
 */
